internal/controllers/bundledeployment/bundledeployment.go

```diff
@@ -198,64 +230,85 @@ func (c *controller) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
-			}
-		}
-		meta.SetStatusCondition(&bd.Status.Conditions, metav1.Condition{
-			Type:    rukpakv1alpha1.TypeHasValidBundle,
-			Status:  status,
-			Reason:  reason,
-			Message: message,
-		})
+
+	if finalizerResult.Updated {
+		// The only thing outside the status that should ever change when handling finalizers
+		// is the list of finalizers in the object's metadata. In particular, we'd expect
+		// finalizers to be added or removed.
+		bd.ObjectMeta.Finalizers = finalizerBundleDelpoyment.ObjectMeta.Finalizers
+	}
+	if finalizerResult.StatusUpdated {
+		bd.Status = finalizerBundleDelpoyment.Status
+	}
+	if finalizerResult.Updated || finalizerResult.StatusUpdated || !bd.GetDeletionTimestamp().IsZero() {
 		return ctrl.Result{}, nil
 	}
 
-	meta.SetStatusCondition(&bd.Status.Conditions, metav1.Condition{
-		Type:    rukpakv1alpha1.TypeHasValidBundle,
-		Status:  metav1.ConditionTrue,
-		Reason:  rukpakv1alpha1.ReasonUnpackSuccessful,
-		Message: fmt.Sprintf("Successfully unpacked the %s Bundle", bundle.GetName()),
-	})
+	unpackResult, err := c.unpacker.Unpack(ctx, bd)
+	if err != nil {
+		return ctrl.Result{}, updateStatusUnpackFailing(&bd.Status, fmt.Errorf("source bundle content: %v", err))
+	}
 
-	bundleFS, err := c.storage.Load(ctx, bundle)
+	switch unpackResult.State {
+	case unpackersource.StatePending:
+		updateStatusUnpackPending(&bd.Status, unpackResult)
+		// There must a limit to number of retries if status is stuck at
+		// unpack pending.
+		return ctrl.Result{}, nil
+	case unpackersource.StateUnpacking:
+		updateStatusUnpacking(&bd.Status, unpackResult)
+		return ctrl.Result{}, nil
+	case unpackersource.StateUnpacked:
+		if err := c.storage.Store(ctx, bd, unpackResult.Bundle); err != nil {
+			return ctrl.Result{}, updateStatusUnpackFailing(&bd.Status, fmt.Errorf("persist bundle content: %v", err))
+		}
+		contentURL, err := c.storage.URLFor(ctx, bd)
+		if err != nil {
+			return ctrl.Result{}, updateStatusUnpackFailing(&bd.Status, fmt.Errorf("get content URL: %v", err))
+		}
+		updateStatusUnpacked(&bd.Status, unpackResult, contentURL)
+	default:
+		return ctrl.Result{}, updateStatusUnpackFailing(&bd.Status, fmt.Errorf("unknown unpack state %q: %v", unpackResult.State, err))
+	}
+
+	bundleFS, err := c.storage.Load(ctx, bd)
 	if err != nil {
 		meta.SetStatusCondition(&bd.Status.Conditions, metav1.Condition{
-			Type:    rukpakv1alpha1.TypeHasValidBundle,
+			Type:    rukpakv1alpha2.TypeHasValidBundle,
 			Status:  metav1.ConditionFalse,
-			Reason:  rukpakv1alpha1.ReasonBundleLoadFailed,
+			Reason:  rukpakv1alpha2.ReasonBundleLoadFailed,
 			Message: err.Error(),
 		})
 		return ctrl.Result{}, err
 	}
 
 	chrt, values, err := c.handler.Handle(ctx, bundleFS, bd)
 	if err != nil {
+		// Should be a better reason than "TypeHasValidBundle", probably "TypeHasPorcessedBundle"?
+		// Cannot use HasUnpacked condition again, as it is being set to true when the contents
+		// are fetched from source and stored in Fs.
```

Comment:
What if we just fold this into the "install" process from the status conditions standpoint and say that if `Handle` fails, that means installation failed?

We can always come back and add another condition later to further disambiguate if we need to.